### ISSUE TITLE
PostJob deferring forever

### ISSUE THREAD

**Issue by belforte** [dmwm/CRABServer#5318]
seems there's no limit to how much and hjow long a PJ can keep deferring itself.
This one e.g. has been doing it for a week.
```
[root@vocms0120 cluster61250.proc0.subproc0]# grep Defer /etc/condor/condor_local/spool/3612/0/cluster33612.proc0.subproc0/postjob.980.3.txt |head -2
Fri, 18 Nov 2016 11:07:31 CET(+0000):INFO:PostJob Defering the execution of the post-job.
Fri, 18 Nov 2016 11:37:37 CET(+0000):INFO:PostJob Defering the execution of the post-job.
[root@vocms0120 cluster61250.proc0.subproc0]# grep Defer /etc/condor/condor_local/spool/3612/0/cluster33612.proc0.subproc0/postjob.980.3.txt |tail -2
Fri, 25 Nov 2016 17:45:43 CET(+0000):INFO:PostJob Defering the execution of the post-job.
Fri, 25 Nov 2016 18:16:11 CET(+0000):INFO:PostJob Defering the execution of the post-job.
[root@vocms0120 cluster61250.proc0.subproc0]# 
```

surely there are good reasons why keeps failing, but some overall control that stop things after N tries, is needed.

e.g. I found another that was failing to get info due to expired proxy in the spool dir.

**Comment by belforte** [comment]
e.g.
```
Fri, 25 Nov 2016 18:16:11 CET(+0000):DEBUG:PostJob Querying ASO view.
Fri, 25 Nov 2016 18:16:11 CET(+0000):ERROR:PostJob Error while querying the RDBMS database.
Traceback (most recent call last):
  File "TaskWorker/Actions/PostJob.py", line 853, in get_transfers_statuses
    view_results = self.couch_database.loadView('AsyncTransfer', 'JobsIdsStatesByWorkflow', query)['rows']
  File "WMCore/Database/CMSCouch.py", line 488, in loadView
    (self.name, design, view), encodedOptions)
  File "WMCore/Services/Requests.py", line 104, in get
    encode, decode, contentType)
  File "WMCore/Database/CMSCouch.py", line 120, in makeRequest
    encode, decode, contentType)
  File "WMCore/Services/Requests.py", line 140, in makeRequest
    encoder, decoder, contentType)
  File "WMCore/Services/Requests.py", line 236, in makeRequest_httplib
    body=encoded_data, headers=headers)
  File "httplib2/__init__.py", line 1436, in request
    (response, content) = self._request(conn, authority, uri, request_uri, method, body, headers, redirections, cachekey)
  File "httplib2/__init__.py", line 1188, in _request
    (response, content) = self._conn_request(conn, request_uri, method, body, headers)
  File "httplib2/__init__.py", line 1123, in _conn_request
    conn.connect()
  File "httplib2/__init__.py", line 911, in connect
    raise SSLHandshakeError(e)
SSLHandshakeError: [Errno 1] _ssl.c:492: error:14094415:SSL routines:SSL3_READ_BYTES:sslv3 alert certificate expired
Fri, 25 Nov 2016 18:16:11 CET(+0000):DEBUG:PostJob Querying transfers statuses using fallback method (i.e. loading each document).
Fri, 25 Nov 2016 18:16:11 CET(+0000):ERROR:PostJob Error retrieving document from ASO database for ID 17e8550394f59eccc4e20c1d6a446efa77fccb048a43d38f574b225d
Traceback (most recent call last):
  File "TaskWorker/Actions/PostJob.py", line 946, in load_transfer_document
    doc = self.couch_database.document(doc_id)
  File "WMCore/Database/CMSCouch.py", line 303, in document
    return Document(id=id, inputDict=self.get(uri))
  File "WMCore/Services/Requests.py", line 104, in get
    encode, decode, contentType)
  File "WMCore/Database/CMSCouch.py", line 120, in makeRequest
    encode, decode, contentType)
  File "WMCore/Services/Requests.py", line 140, in makeRequest
    encoder, decoder, contentType)
  File "WMCore/Services/Requests.py", line 236, in makeRequest_httplib
    body=encoded_data, headers=headers)
  File "httplib2/__init__.py", line 1436, in request
    (response, content) = self._request(conn, authority, uri, request_uri, method, body, headers, redirections, cachekey)
  File "httplib2/__init__.py", line 1188, in _request
    (response, content) = self._conn_request(conn, request_uri, method, body, headers)
  File "httplib2/__init__.py", line 1123, in _conn_request
    conn.connect()
  File "httplib2/__init__.py", line 911, in connect
    raise SSLHandshakeError(e)
SSLHandshakeError: [Errno 1] _ssl.c:492: error:14094415:SSL routines:SSL3_READ_BYTES:sslv3 alert certificate expired
Fri, 25 Nov 2016 18:16:11 CET(+0000):INFO:PostJob Got statuses: unknown; 0.0 hours since transfer submit.
Fri, 25 Nov 2016 18:16:11 CET(+0000):INFO:PostJob Defering the execution of the post-job.
Ended TaskManagerBootstrap with code 4

```

**Comment by mmascher** [comment]
There is a timeout of one day for the postjob (configurable in the TW configuration). The one day is because we want to wait for the transfer to finish. BTW, was this a task submitted before or after Thursday (when we applied the patch)? Do you have the taskname?

I am more interested in knowing why the proxy expired and why the RenewRemoteProxy did not work.

**Comment by belforte** [comment]
One day is OK. But the PH in the example had been running for 7 days. I think in a case like this the code does not reach the place whiere it checks the day.
About the proxy, just a guess, but could it be that proxy renewal is not not on schedd's in drain i.e. not in listed in the configuration ?

**Comment by belforte** [comment]
One day is OK. But the PH in the example had been running for 7 days. I think in a case like this the code does not reach the place whiere it checks the day.
About the proxy, just a guess, but could it be that proxy renewal is not not on schedd's in drain i.e. not in listed in the configuration ?

**Comment by belforte** [comment]
@mmascher has found why the one-day timeout was not obyed. It can happen when initial posting to Couch for ASo fails on the WN, then the PJ makes it but does nor correctly record the time it did, so the check for "one day for xfer" keeps finding the tale telling 
```
INFO:PostJob Got statuses: unknown; 0.0 hours since transfer submit.
```
and PJ goes on .. sort of forever (I guess until some other things kills the task, like forced disk cleanup)

This happens even if proxy was properly renewed, but when I looked at vocms120 which is in drain I also found the expired proxy and that is indeed due to the fact that proxy renewal procedure does not care for schedd's in drain. That's not optimal, but in the end it means simly that on a draining schedd tasks die after one week, which is acceptable.

Marco says he has put in a patch already so that PJ correctly record the ASO injection date,
